internal/reconciler/implementation.go

```diff
@@ -0,0 +1,109 @@
+package reconciler
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"github.com/nginxinc/nginx-kubernetes-gateway/internal/events"
+)
+
+// NamespacedNameFilterFunc is a function that returns true if the resource should be processed by the reconciler.
+// If the function returns false, the reconciler will log the returned string.
+type NamespacedNameFilterFunc func(nsname types.NamespacedName) (bool, string)
+
+// Config contains the configuration for the Implementation.
+type Config struct {
+	// Getter gets a resource from the k8s API.
+	Getter Getter
+	// ObjectType is the type of the resource that the reconciler will reconcile.
+	ObjectType client.Object
+	// EventCh is the channel where the reconciler will send events.
+	EventCh chan<- interface{}
+	// NamespacedNameFilter filters resources the controller will process. Can be nil.
+	NamespacedNameFilter NamespacedNameFilterFunc
+}
+
+// Implementation is a reconciler for Kubernetes resources.
+// It implements the reconcile.Reconciler interface.
+// A successful reconciliation of a resource has the two possible outcomes:
+// (1) If the resource is deleted, the Implementation will send a DeleteEvent to the event channel.
+// (2) If the resource is upserted (created or updated), the Implementation will send an UpsertEvent
+// to the event channel.
+type Implementation struct {
+	cfg Config
+}
+
+var _ reconcile.Reconciler = &Implementation{}
+
+// NewImplementation creates a new Implementation.
+func NewImplementation(cfg Config) *Implementation {
+	return &Implementation{
+		cfg: cfg,
+	}
+}
+
+func newObject(objectType client.Object) client.Object {
+	// without Elem(), t will be a pointer to the type. For example, *v1beta1.Gateway, not v1beta1.Gateway
+	t := reflect.TypeOf(objectType).Elem()
+
+	return reflect.New(t).Interface().(client.Object)
```

Comment:
A `client.Object` is a `runtime.Object` and has a `DeepCopyObject`. Have you tried that here instead of `reflect`?

Reply:
DeepCopyObject looks easier. I think I disregarded it initially it because of assuming deep copying would be more expensive than using reflection, but I didn't check it though. I don't see any big difference.

```
func newObject2(objectType client.Object) client.Object {
	return objectType.DeepCopyObject().(client.Object)
}
```

```
func BenchmarkNewObject(b *testing.B) {
	objectType := &v1beta1.HTTPRoute{}
	for i := 0; i < b.N; i++ {
		_ = newObject(objectType)
	}
}

func BenchmarkNewObject2(b *testing.B) {
	objectType := &v1beta1.HTTPRoute{}
	for i := 0; i < b.N; i++ {
		_ = newObject2(objectType)
	}
}
```

```
BenchmarkNewObject-4             6310166               201.0 ns/op
BenchmarkNewObject2-4            4395318               263.1 ns/op
```